### PR TITLE
Cyclops sync fix and improvements

### PR DESCRIPTION
**Pull request by tornac1234**
This PR will try to fix the existing bugs and maybe add some things that weren't synced yet

- [x] Fix cyclops sonar executing from everywhere [Issue#1493](https://github.com/SubnauticaNitrox/Nitrox/issues/1493)
- [x] Fix motor state not syncing
- [x] Fix vehicle loading errors about colors packets (silent errors in the logs but don't do anything wrong)
- [x] Fix motor state not restoring when spawning far away from the cyclops
- [x] Fix sonar button state not being synced
- [x] Fix spawning far away of the Cyclops when relogging (has to do with wrong calculations of the client position sent to the server)

### CODE REVIEW

**Comment by tornac1234** [comment]
Does the last commit fixes [Issue#1465](https://github.com/SubnauticaNitrox/Nitrox/issues/1465) ?

File: NitroxPatcher/Patches/Dynamic/CyclopsSonarButton_Update_Patch.cs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+using System.Reflection.Emit;
+using HarmonyLib;
+using NitroxClient.GameLogic;
+using NitroxClient.MonoBehaviours;
+using NitroxModel.DataStructures;
+using NitroxModel.Helper;
+
+namespace NitroxPatcher.Patches.Dynamic
+{
+    public class CyclopsSonarButton_Update_Patch : NitroxPatch, IDynamicPatch
```
Comment (Measurity) [inline NitroxPatcher/Patches/Dynamic/CyclopsSonarButton_Update_Patch.cs:12]: Do you know which logic is causing the sonar ping to not be visible when you exit the cyclops while another pilot is still driving?

What about adding a comment here? Something like this?
```suggestion
    /// <summary>
    /// Prevents sonar from turning off automatically when the player stops piloting the Cyclops.
    /// The sonar will stay on until the player leaves the vehicle and automatically turns on when they enter again (if sonar was on at that time). 
    /// </summary>
    public class CyclopsSonarButton_Update_Patch : NitroxPatch, IDynamicPatch
```

**Review by Measurity — APPROVED** [review]
Tested with multiple instances and couldn't find an issue.

File: NitroxPatcher/Patches/Dynamic/CyclopsSonarButton_SonarPing_Patch.cs [hunk 2]
```diff
@@ -75,7 +77,7 @@ public static IEnumerable<CodeInstruction> Transpiler(MethodBase original, IEnum
                  */
                 if (instruction.opcode.Equals(OpCodes.Brtrue))
                 {
-                    if (instructionList[i - 1].opcode.Equals(OpCodes.Ldloc_1) && instructionList[i + 1].opcode.Equals(OpCodes.Ldarg_0))
+                    if (instructionList[i - 1].opcode.Equals(OpCodes.Call) && instructionList[i + 1].opcode.Equals(OpCodes.Ldarg_0))
```
[the remark this answers is not in the document]
Comment (tornac1234) [reply NitroxPatcher/Patches/Dynamic/CyclopsSonarButton_SonarPing_Patch.cs:80]: As energy consumption sync seems weird for now, I don't think we should make changes about this. Also, when receiving the sonar ping packet, we should still simulate the energy loss as the cyclops did use the sonar ping for another client

File: NitroxPatcher/Patches/Dynamic/CyclopsSonarButton_Update_Patch.cs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+            CodeInstruction brInstruction = new(OpCodes.Brfalse);
+            for (int i = 0; i < codeInstructions.Count; i++)
+            {
+                CodeInstruction instruction = codeInstructions[i];
+
+                if (instruction.opcode.Equals(INJECTION_OPCODE) && instruction.operand.Equals(INJECTION_OPERAND))
+                {
+                    // The second line after the current instruction should be a Brtrue, we need its operand to have the same jump label for our brfalse
+                    CodeInstruction nextBr = codeInstructions[i + 2];
+                    Validate.IsTrue(nextBr.opcode.Equals(OpCodes.Brtrue), "Looks like subnautica code has changed. Update jump offset!");
```
Comment (Jannify) [inline NitroxPatcher/Patches/Dynamic/CyclopsSonarButton_Update_Patch.cs:52]: This (constant validation statements in general (like at the start of the transpiler)) should be checked in the test and not at runtime.

File: NitroxPatcher/Patches/Dynamic/CyclopsSonarButton_Update_Patch.cs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            
+            /* Normally in the Update()
+             * if (Player.main.GetMode() == Player.Mode.Normal && this.sonarActive)
+		     * {
+		     *     this.TurnOffSonar();
+		     * }
+             * this part will be changed into:
+             * if (CyclopsSonarButton_Update_Patch.ShouldTurnOff() && Player.main.GetMode() == Player.Mode.Normal && this.sonarActive)
+             */
+            List<CodeInstruction> codeInstructions = new List<CodeInstruction>(instructions);
```
Comment (Jannify) [inline NitroxPatcher/Patches/Dynamic/CyclopsSonarButton_Update_Patch.cs:41]: Is there a reason you casting to List? Afaik you van iterate a IEnumerable with foreach.

**Review by Jannify — CHANGES_REQUESTED** [review]

File: NitroxPatcher/Patches/Dynamic/CyclopsSonarButton_Update_Patch.cs (hunk 4 above)
Comment (tornac1234) [reply NitroxPatcher/Patches/Dynamic/CyclopsSonarButton_Update_Patch.cs:41]: There's a moment I need to get the code instruction that's 2 lines further, that's why I don't use a foreach

File: NitroxPatcher/Patches/Dynamic/CyclopsSonarButton_Update_Patch.cs [hunk 5]
```diff
@@ -16,8 +16,8 @@ public class CyclopsSonarButton_Update_Patch : NitroxPatch, IDynamicPatch
     {
         public static readonly MethodInfo TARGET_METHOD = Reflect.Method((CyclopsSonarButton t) => t.Update());
 
-        private static readonly OpCode INJECTION_OPCODE = OpCodes.Ldsfld;
-        private static readonly object INJECTION_OPERAND = Reflect.Field(() => Player.main);
+        public static readonly OpCode INJECTION_OPCODE = OpCodes.Ldsfld;
+        public static readonly object INJECTION_OPERAND = Reflect.Field(() => Player.main);
```
Comment (Measurity) [inline NitroxPatcher/Patches/Dynamic/CyclopsSonarButton_Update_Patch.cs:20]: `internal` is preferred. The Nitrox.Test project has exclusive access to all `internal` members for this reason.

File: Nitrox.Test/Patcher/Patches/CyclopsSonarButton_Update_PatchTest.cs [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+using System.Linq;
+using System.Reflection.Emit;
+using HarmonyLib;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using NitroxModel.Helper;
+using NitroxPatcher.Patches.Dynamic;
+using NitroxTest.Patcher.Test;
+
+namespace Nitrox.Test.Patcher.Patches
+{
```
Comment (Jannify) [inline Nitrox.Test/Patcher/Patches/CyclopsSonarButton_Update_PatchTest.cs:12]:
```suggestion
namespace NitroxPatcher.Patches.Dynamic;
```

File: NitroxPatcher/Patches/Dynamic/CyclopsSonarButton_Update_Patch.cs [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+        public static void Prefix(CyclopsSonarButton __instance)
+        {
+            currentCyclopsId = NitroxEntity.GetId(__instance.subRoot.gameObject);
+        }
+
+        public static IEnumerable<CodeInstruction> Transpiler(MethodBase methodBase, IEnumerable<CodeInstruction> instructions)
+        {
+            Validate.NotNull(INJECTION_OPERAND);
+
```
Comment (Jannify) [inline NitroxPatcher/Patches/Dynamic/CyclopsSonarButton_Update_Patch.cs:32]:
```suggestion
```

**Review by Jannify — CHANGES_REQUESTED** [review]

**Pull request merged**